src/main/scala/com/kinja/amqp/AmqpProducer.scala

```diff
@@ -16,24 +26,59 @@ import play.api.libs.json._
 
 import java.util.concurrent.TimeUnit
 
+import scala.concurrent.Future
+import scala.concurrent.duration._
+import scala.util.Try
+
 class AmqpProducer(
 	connection: ActorRef,
 	actorSystem: ActorSystem,
+	messageStore: MessageStore,
 	connectionTimeOut: Long,
+	askTimeout: Long,
 	logger: Slf4jLogger)(exchange: ExchangeParameters) {
 
 	private val channel: ActorRef = createChannel()
 
 	Amqp.waitForConnection(actorSystem, connection, channel).await(connectionTimeOut, TimeUnit.SECONDS)
 
-	def publish[A: Writes](routingKey: String, message: A): Unit = {
+	channel ! ConfirmSelect
+	channel ! AddConfirmListener(createConfirmListener)
+
+	def publish[A: Writes](routingKey: String, message: A, saveTimeMillis: Long = System.currentTimeMillis()): Future[Try[Unit]] = {
 		val json = Json.toJson(message)
 		val bytes = json.toString.getBytes(java.nio.charset.Charset.forName("UTF-8"))
 		val properties = new BasicProperties.Builder().deliveryMode(2).build()
-		channel ! Publish(exchange.name, routingKey, bytes, properties = Some(properties), mandatory = true, immediate = false)
+		implicit val timeout = Timeout(askTimeout.seconds)
+		import scala.concurrent.ExecutionContext.Implicits.global
```

Comment:
instead of this, publish must take an implicit ExecutionContext parameter. that is the conventional pattern of injecting execution context as a dependency